`cmake`: dependencies between `CMake` targets work only if the dependent target is imported

`subprojects/cmMod`:
```
cmake_minimum_required(VERSION 3.5)

project(cmMod)
set (CMAKE_CXX_STANDARD 14)

include_directories(${CMAKE_CURRENT_BINARY_DIR})

add_library(cmModLib1 STATIC cmMod1.cpp)
add_library(cmModLib2 STATIC cmMod2.cpp)
# This is what is not working
target_link_libraries(cmModLib1 PUBLIC cmModLib2)
```

`meson.build`:
```
project('cmakeSubTest', ['c', 'cpp'])

cm = import('cmake')

sub_pro = cm.subproject('cmMod')
# require only the first one
sub_dep = sub_pro.dependency('cmModLib1', include_type: 'system')

exe1 = executable('main', ['main.cpp'], dependencies: [sub_dep])
test('test1', exe1)
```

Result:
```
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -v
[1/4] /usr/local/bin/g++ -Isubprojects/cmMod/libcmModLib1.a.p -Isubprojects/cmMod -I../subprojects/cmMod -Isubprojects/cmMod/__CMake_build -I../subprojects/cmMod/__CMake_build -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu++14 -O0 -g -g -DcmModClass=cmModClass1 -MD -MQ subprojects/cmMod/libcmModLib1.a.p/cmMod.cpp.o -MF subprojects/cmMod/libcmModLib1.a.p/cmMod.cpp.o.d -o subprojects/cmMod/libcmModLib1.a.p/cmMod.cpp.o -c ../subprojects/cmMod/cmMod.cpp
[2/4] /usr/local/bin/g++ -Imain.p -I. -I.. -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -isystem../subprojects/cmMod -isystemsubprojects/cmMod -isystem../subprojects/cmMod/__CMake_build -isystemsubprojects/cmMod/__CMake_build -MD -MQ main.p/main.cpp.o -MF main.p/main.cpp.o.d -o main.p/main.cpp.o -c ../main.cpp
[3/4] rm -f subprojects/cmMod/libcmModLib1.a && gcc-ar csrDT subprojects/cmMod/libcmModLib1.a subprojects/cmMod/libcmModLib1.a.p/cmMod.cpp.o
[4/4] /usr/local/bin/g++  -o main main.p/main.cpp.o -Wl,--as-needed -Wl,--no-undefined subprojects/cmMod/libcmModLib1.a
FAILED: main 
/usr/local/bin/g++  -o main main.p/main.cpp.o -Wl,--as-needed -Wl,--no-undefined subprojects/cmMod/libcmModLib1.a
/usr/bin/ld: main.p/main.cpp.o: in function `main':
/home/mmom/src/meson/test cases/cmake/29 cascading deps/build/../main.cpp:14: undefined reference to `cmModClass2::cmModClass2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
/usr/bin/ld: /home/mmom/src/meson/test cases/cmake/29 cascading deps/build/../main.cpp:16: undefined reference to `cmModClass2::getStr[abi:cxx11]() const'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

The problem is that when resolving the dependencies, only targets that have `IMPORTED_LOCATION` or `IMPORTED_IMPLIB` result in an actual library dependency in `meson`:

https://github.com/mmomtchev/meson/blob/648a3ddd65a6010dbfaf42bf9de624c2ae47bccd/mesonbuild/cmake/tracetargets.py#L118

Only imported targets have these, internal `CMake` targets have to be included via `LINK_LIBRARIES` / `INTERFACE_LINK_LIBRARIES` - in the code, these are used only for recursion.

--- a/mesonbuild/cmake/interpreter.py
+++ b/mesonbuild/cmake/interpreter.py
@@ -223,6 +223,7 @@ class ConverterTarget:
         self.install = target.install
         self.install_dir: T.Optional[Path] = None
         self.link_libraries = target.link_libraries
+        self.link_targets: T.List[str] = []
         self.link_flags = target.link_flags + target.link_lang_flags
         self.public_link_flags: T.List[str] = []
         self.depends_raw: T.List[str] = []
@@ -363,6 +364,8 @@ class ConverterTarget:
             self.public_link_flags += rtgt.public_link_flags
             self.public_compile_opts += rtgt.public_compile_opts
             self.link_libraries += rtgt.libraries
+            self.depends_raw += rtgt.target_dependencies
+            self.link_targets += rtgt.target_dependencies
 
         elif self.type.upper() not in ['EXECUTABLE', 'OBJECT_LIBRARY']:
             mlog.warning('CMake: Target', mlog.bold(self.cmake_name), 'not found in CMake trace. This can lead to build errors')
@@ -957,17 +960,27 @@ class CMakeInterpreter:
                 object_libs += [tgt]
             self.languages += [x for x in tgt.languages if x not in self.languages]
 
-        # Second pass: Detect object library dependencies
+        # Second pass: Populate link_with project internal targets
+        for tgt in self.targets:
+            for i in tgt.link_targets:
+                # Handle target-based link libraries
+                link_with = self.output_target_map.target(i)
+                if not link_with or isinstance(link_with, ConverterCustomTarget):
+                    # Generated file etc.
+                    continue
+                tgt.link_with.append(link_with)
+
+        # Third pass: Detect object library dependencies
         for tgt in self.targets:
             tgt.process_object_libs(object_libs, self._object_lib_workaround)
 
-        # Third pass: Reassign dependencies to avoid some loops
+        # Fourth pass: Reassign dependencies to avoid some loops
         for tgt in self.targets:
             tgt.process_inter_target_dependencies()
         for ctgt in self.custom_targets:
             ctgt.process_inter_target_dependencies()
 
-        # Fourth pass: Remove rassigned dependencies
+        # Fifth pass: Remove reassigned dependencies
         for tgt in self.targets:
             tgt.cleanup_dependencies()
 

--- a/mesonbuild/cmake/tracetargets.py
+++ b/mesonbuild/cmake/tracetargets.py
@@ -45,6 +45,7 @@ class ResolvedTarget:
         self.public_link_flags:   T.List[str] = []
         self.public_compile_opts: T.List[str] = []
         self.libraries:           T.List[str] = []
+        self.target_dependencies: T.List[str] = []
 
 def resolve_cmake_trace_targets(target_name: str,
                                 trace: 'CMakeTraceParser',
@@ -144,9 +145,13 @@ def resolve_cmake_trace_targets(target_name: str,
             targets += [x for x in tgt.properties['IMPORTED_LOCATION'] if x]
 
         if 'LINK_LIBRARIES' in tgt.properties:
-            targets += [x for x in tgt.properties['LINK_LIBRARIES'] if x]
+            link_libraries = [x for x in tgt.properties['LINK_LIBRARIES'] if x]
+            targets += link_libraries
+            res.target_dependencies += link_libraries
         if 'INTERFACE_LINK_LIBRARIES' in tgt.properties:
-            targets += [x for x in tgt.properties['INTERFACE_LINK_LIBRARIES'] if x]
+            link_libraries = [x for x in tgt.properties['INTERFACE_LINK_LIBRARIES'] if x]
+            targets += link_libraries
+            res.target_dependencies += link_libraries
 
         if f'IMPORTED_LINK_DEPENDENT_LIBRARIES_{cfg}' in tgt.properties:
             targets += [x for x in tgt.properties[f'IMPORTED_LINK_DEPENDENT_LIBRARIES_{cfg}'] if x]

